File: go/test/hook_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gitlab-sdk/go"
	"github.com/voxgig-sdk/gitlab-sdk/go/core"

	vs "github.com/voxgig-sdk/gitlab-sdk/go/utility/struct"
)

func TestHookEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Hook(nil)
		if ent == nil {
			t.Fatal("expected non-nil HookEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := hookBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "hook." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_HOOK_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		hookRef01Ent := client.Hook(nil)
		hookRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "hook"}, setup.data), "hook_ref01"))
		hookRef01Data["hook_id"] = setup.idmap["hook01"]

		hookRef01DataResult, err := hookRef01Ent.Create(hookRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		hookRef01Data = core.ToMapAny(hookRef01DataResult)
		if hookRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// UPDATE
		hookRef01DataUp0Up := map[string]any{
		}

		hookRef01ResdataUp0Result, err := hookRef01Ent.Update(hookRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		hookRef01ResdataUp0 := core.ToMapAny(hookRef01ResdataUp0Result)
		if hookRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}

		// REMOVE
		hookRef01MatchRm0 := map[string]any{
			"id": hookRef01Data["id"],
		}
		_, err = hookRef01Ent.Remove(hookRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

	})
}

func hookBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "hook", "HookTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read hook test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse hook test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"hook01", "hook02", "hook03", "custom_header01", "custom_header02", "custom_header03", "url_variable01", "url_variable02", "url_variable03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITLAB_TEST_HOOK_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITLAB_TEST_HOOK_ENTID": idmap,
		"GITLAB_TEST_LIVE":      "FALSE",
		"GITLAB_TEST_EXPLAIN":   "FALSE",
		"GITLAB_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GITLAB_TEST_HOOK_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITLAB_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GITLAB_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGitlabSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITLAB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITLAB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/hook_entity_test.lua
-- Hook entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("gitlab_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("HookEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Hook(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = hook_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "update", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "hook." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_HOOK_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local hook_ref01_ent = client:Hook(nil)
    local hook_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.hook"), "hook_ref01"))
    hook_ref01_data["hook_id"] = setup.idmap["hook01"]

    local hook_ref01_data_result, err = hook_ref01_ent:create(hook_ref01_data, nil)
    assert.is_nil(err)
    hook_ref01_data = helpers.to_map(hook_ref01_data_result)
    assert.is_not_nil(hook_ref01_data)

    -- UPDATE
    local hook_ref01_data_up0_up = {
    }

    local hook_ref01_resdata_up0_result, err = hook_ref01_ent:update(hook_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local hook_ref01_resdata_up0 = helpers.to_map(hook_ref01_resdata_up0_result)
    assert.is_not_nil(hook_ref01_resdata_up0)

    -- REMOVE
    local hook_ref01_match_rm0 = {
      id = hook_ref01_data["id"],
    }
    local _, err = hook_ref01_ent:remove(hook_ref01_match_rm0, nil)
    assert.is_nil(err)

  end)
end)

function hook_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/hook/HookTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read hook test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "hook01", "hook02", "hook03", "custom_header01", "custom_header02", "custom_header03", "url_variable01", "url_variable02", "url_variable03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITLAB_TEST_HOOK_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITLAB_TEST_HOOK_ENTID"] = idmap,
    ["GITLAB_TEST_LIVE"] = "FALSE",
    ["GITLAB_TEST_EXPLAIN"] = "FALSE",
    ["GITLAB_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["GITLAB_TEST_HOOK_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["GITLAB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["GITLAB_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITLAB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITLAB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
